### DISCUSSION TITLE
Having problems with mocking Zustand 4.2.0 in Jest

### DISCUSSION THREAD

**Discussion by mgrahamspoton** [pmndrs/zustand#1508] [Q&A]
Hello,
I have recently introduced Zustand to an application we are working on. We love it so far.
We are having some challenges getting Zustand integrated properly with Jest, even after following the docs.

We are on:
- "jest": "^26.6.3",
- "zustand": "^4.2.0"

we get the following error when our vanilla.js mock file has:
```
const create = f =>
  f === undefined ? createImpl : createImpl(f)

// when creating a store, we get its initial state, create a reset function and add it in the set
const createImpl = (createState) => {
  const store = actualCreate(createState)
  const initialState = store.getState()
  storeResetFns.add(() => store.setState(initialState, true))
  return store
}
```
![image](https://user-images.githubusercontent.com/89858620/210645730-4204699c-6b22-460e-a90a-d74aa701382f.png)

if we use:
```
const create = () => createState => {
    const store = actualCreate(createState);
    const initialState = store.getState();
    storeResetFns.add(() => store.setState(initialState, true));
    return store;
};
```

we get the following error:
![image](https://user-images.githubusercontent.com/89858620/210646283-d2169225-6c4e-451d-a717-7d7daa28a861.png)

**Comment by dai-shi** [comment]
Hi, first thing to confirm, is this problem specific to v4.2.0? Can you try older versions if they fixes?

Then, can you reproduce the problem with https://codesandbox.io/s/react-new ?

**Comment by dai-shi** [comment]
```js
const create = f =>
  f === undefined ? createImpl : createImpl(f)
```
Do we have that somewhere in docs?
I think we should follow the new pattern.
```js
const create = (createState) => {
  if (!createState) return create;
  const store = actualCreate(createState);
    ...
```
@dbritto-dev what do you think?

**Reply by dbritto-dev** [reply]
I don't think so but we shall add it and update the default readme as well

**Reply by dai-shi** [reply]
https://github.com/pmndrs/zustand/blob/0ef5d1e81be9c953c20b60aed416a7662bf31d89/docs/guides/testing.mdx?plain=1#L53-L61

Hm, actually it's not like what I expected.

Oh, it wasn't the testing guide.
It's in the reset guide.
https://github.com/pmndrs/zustand/blob/0ef5d1e81be9c953c20b60aed416a7662bf31d89/docs/guides/how-to-reset-state.md?plain=1#L54-L62

What do you think about following the new pattern?

> update the default readme as well

Do we have something in readme?

**Reply by dbritto-dev** [reply]
Yeah, we need to update the docs to follow the new pattern including the default readme (`README.md`)

**Reply by dbritto-dev** [reply]
Just take a look of this https://github.com/pmndrs/zustand/blob/main/readme.md it's outdated as well

**Reply by dai-shi** [reply]
We don't have mocking guide or reset guide in readme. If you are talking about `create(...)` style, it's intentional for now. See: #1430 

> we need to update the docs to follow the new pattern

Feel free to work on docs, when you have time.

**Reply by dbritto-dev** [reply]
But we have the immer implementation in the readme, so we need to update that. I think we need to update the docs to follow the new pattern instead of keeping the old one because that could lead to some issues like he have on this issue. What do you think?

BTW, I can start working on the docs

**Reply by dai-shi** [reply]
> But we have the immer implementation in the readme, so we need to update that. I think we need to update the docs to follow the new pattern instead of keeping the old one because that could lead to some issues like he have on this issue. What do you think?

I'm not sure if I follow. What update are you suggesting?

> BTW, I can start working on the docs

👍

**Reply by dbritto-dev** [reply]
I was talking about this part

![IMG_20230106_004004.jpg](https://user-images.githubusercontent.com/1697714/210937715-993e077b-cea3-4e48-9030-8939d20c45d7.jpg)

**Reply by dai-shi** [reply]
That part should be working fine in JS, right?
For now, the readme is entirely in JS, not in TS, which is controversial though.

**Reply by dbritto-dev** [reply]
Sounds good I will work on the other docs
